### PR TITLE
Adding AttributeMeta, StringMeta and MapMeta

### PR DESCRIPTION
**Pull request by GDYendell**
How is this?

### CODE REVIEW

**Comment by GDYendell** [comment]
Is the StringMeta validate() too restrictive?

File: malcolm/core/mapmeta.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+
+class MapMeta(Loggable):
+    """A meta object to store a set of attribute metas"""
+
+    def __init__(self, name):
+        super(MapMeta, self).__init__(logger_name=name)
+
+        self.name = name
+        self.elements = OrderedDict({})
```
Comment (thomascobb) [inline malcolm/core/mapmeta.py]: You don't need to pass {} to OrderedDict

File: malcolm/core/mapmeta.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    def add_element(self, attribute_meta, required=False):
+        """
+        Add an element, stating whether it is required.
+
+        Args:
+            attribute_meta(AttributeMeta): Attribute instance to store
+            required(bool): Whether attribute is required or optional
+        """
+
+        self.elements[attribute_meta.name] = (attribute_meta, required)
```
Comment (thomascobb) [inline malcolm/core/mapmeta.py]: Can you make the self.required a list of strings so it's easy to serialise?

**Comment by thomascobb** [comment]
I was a little more permissive here:
https://github.com/dls-controls/malcolm/blob/master/malcolm/core/vtype.py#L179

Just less lines of code, don't really mind either way

File: tests/test_core/test_attributemeta.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+    def setUp(self):
+        self.attribute_meta = AttributeMeta("Test")
+
+    def test_given_validate_called_then_raise_error(self):
+
+        expected_error_message = \
+            "Abstract validate function must be implemented in child classes"
+
+        with self.assertRaises(NotImplementedError) as error:
```
Comment (thomascobb) [inline tests/test_core/test_attributemeta.py]: Didn't know you could do this!

File: tests/test_core/test_stringmeta.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+        self.assertEqual("15", response)
+
+    def test_given_value_float_then_cast_and_set(self):
+        response = self.string_meta.validate(12.8)
+
+        self.assertEqual("12.8", response)
+
+    def test_given_value_dict_then_raise_error(self):
+        expected_error_message = "Value must be of type str or castable to str"
```
Comment (thomascobb) [inline tests/test_core/test_stringmeta.py]: Although checking the error message is probably a bit too much work...

File: malcolm/core/mapmeta.py [hunk 5]
```diff
@@ -5,21 +5,23 @@
...
-        Add an element, stating whether it is required.
+        Add an element and whether it is required.
 
         Args:
             attribute_meta(AttributeMeta): Attribute instance to store
             required(bool): Whether attribute is required or optional
         """
 
-        self.elements[attribute_meta.name] = (attribute_meta, required)
+        self.elements[attribute_meta.name] = attribute_meta
```
Comment (thomascobb) [inline malcolm/core/mapmeta.py]: What happens if we add an attribute_meta which is already in the map?

**Pull request merged**